File: campus-navigate-app/expo-app/src/tiled/model/surfaceModel.ts
import type {
  OverlayPoint,
  SurfaceRect,
  TileBounds,
  TiledMap,
  TiledTileLayer,
} from '../type';

export function calculateChunkTileBounds(layers: TiledTileLayer[]): TileBounds {
  let minX = Infinity;
  let minY = Infinity;
  let maxX = -Infinity;
  let maxY = -Infinity;

  for (const layer of layers) {
    for (const chunk of layer.chunks ?? []) {
      minX = Math.min(minX, chunk.x);
      minY = Math.min(minY, chunk.y);
      maxX = Math.max(maxX, chunk.x + chunk.width);
      maxY = Math.max(maxY, chunk.y + chunk.height);
    }
  }

  if (!Number.isFinite(minX) || !Number.isFinite(minY)) {
    throw new Error('No tile chunks were found for map bounds metadata.');
  }

  return {
    minX,
    minY,
    maxX,
    maxY,
    width: maxX - minX,
    height: maxY - minY,
  };
}

export function createSurface(bounds: TileBounds, map: TiledMap): SurfaceRect {
  return {
    originX: bounds.minX * map.tilewidth,
    originY: bounds.minY * map.tileheight,
    width: bounds.width * map.tilewidth,
    height: bounds.height * map.tileheight,
  };
}

export function worldToScreenPoint(
  point: { x: number; y: number },
  surface: SurfaceRect,
): OverlayPoint {
  return {
    tiledX: point.x,
    tiledY: point.y,
    screenX: point.x - surface.originX,
    screenY: point.y - surface.originY,
  };
}

File: tarumt-nav-app/flutter_app/lib/domain/tiled/map/surface_model.dart
import 'package:indoor_navigation/domain/tiled/tiled_models.dart';

TileBounds calculateChunkTileBounds(List<TiledTileLayer> layers) {
  int? minX;
  int? minY;
  int? maxX;
  int? maxY;

  for (final layer in layers) {
    for (final chunk in layer.chunks ?? const <TiledChunk>[]) {
      minX = minX == null || chunk.x < minX ? chunk.x : minX;
      minY = minY == null || chunk.y < minY ? chunk.y : minY;
      final chunkMaxX = chunk.x + chunk.width;
      final chunkMaxY = chunk.y + chunk.height;
      maxX = maxX == null || chunkMaxX > maxX ? chunkMaxX : maxX;
      maxY = maxY == null || chunkMaxY > maxY ? chunkMaxY : maxY;
    }
  }

  if (minX == null || minY == null || maxX == null || maxY == null) {
    throw StateError('No tile chunks were found for map bounds metadata.');
  }

  return TileBounds(
    minX: minX,
    minY: minY,
    maxX: maxX,
    maxY: maxY,
    width: maxX - minX,
    height: maxY - minY,
  );
}

SurfaceRect createSurface(TileBounds bounds, TiledMap map) {
  return SurfaceRect(
    originX: (bounds.minX * map.tileWidth).toDouble(),
    originY: (bounds.minY * map.tileHeight).toDouble(),
    width: (bounds.width * map.tileWidth).toDouble(),
    height: (bounds.height * map.tileHeight).toDouble(),
  );
}

OverlayPoint worldToScreenPoint({
  required double x,
  required double y,
  required SurfaceRect surface,
}) {
  return OverlayPoint(
    tiledX: x,
    tiledY: y,
    screenX: x - surface.originX,
    screenY: y - surface.originY,
  );
}
